File: clients/android/app/src/main/java/eu/mycellium/android/ClientHolder.kt
package eu.mycellium.android

import android.content.Context
import uniffi.mycellium_sdk.MyceliumClient

/**
 * Process-wide holder for the one [MyceliumClient]. The client is a stateful
 * handle to this account on this device; the app builds exactly one and shares
 * it (the Rust object guards its own interior state with a Mutex, so it is safe
 * to call from any thread).
 *
 * The client is built with [MyceliumClient.newWithSecretStore] — the production
 * constructor — backed by [AndroidKeystoreSecretStore]. The plaintext dev
 * constructor `MyceliumClient(dataDir)` is deliberately never used.
 *
 * [get] BLOCKS: `newWithSecretStore` opens the encrypted store and touches the
 * Keystore. Call it off the main thread (the ViewModel builds it on
 * `Dispatchers.IO`).
 */
object ClientHolder {

    @Volatile
    private var client: MyceliumClient? = null

    fun get(context: Context): MyceliumClient {
        client?.let { return it }
        synchronized(this) {
            client?.let { return it }
            val app = context.applicationContext
            val built = MyceliumClient.newWithSecretStore(
                // data_dir: app-private, sandboxed storage for the encrypted store.
                app.filesDir.path,
                // The OS-keystore-backed secret store for the account root key (#65).
                AndroidKeystoreSecretStore(app),
            )
            client = built
            return built
        }
    }
}

File: clients/apple/App/ClientHolder.swift
// Process-wide holder for the one `MyceliumClient` (issues #68/#69).
//
// Apple-only. This file is part of the Xcode SwiftUI app target, NOT the
// SwiftPM package — it is excluded from `swift build`/`swift test` on Linux.
//
// The client is a stateful handle to this account on this device; the app
// builds exactly one and shares it (the Rust object guards its own interior
// state with a Mutex, so it is safe to call from any thread).
//
// It is built with the PRODUCTION constructor `MyceliumClient.newWithSecretStore`
// backed by `KeychainSecretStore` (issue #65). The plaintext dev constructor
// `MyceliumClient(dataDir:)` is deliberately never used here.

import Foundation
import MyceliumSDK
import MyceliumSecrets

/// Lazily builds and shares the single `MyceliumClient`. `get()` BLOCKS
/// (`newWithSecretStore` opens the encrypted store and touches the Keychain),
/// so call it off the main thread — the view-model does, on a background task.
enum ClientHolder {

    private static let lock = NSLock()
    nonisolated(unsafe) private static var client: MyceliumClient?

    /// The app-private data directory for the encrypted store (Application
    /// Support, which is backed up unless excluded — the SDK's `store` is
    /// keyed from the identity, and the identity itself lives in the Keychain).
    private static var dataDir: String {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let dir = base.appendingPathComponent("Mycellium", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.path
    }

    static func get() throws -> MyceliumClient {
        lock.lock(); defer { lock.unlock() }
        if let client { return client }
        let built = try MyceliumClient.newWithSecretStore(
            dataDir: dataDir,
            // Single-account app -> empty namespace. Give a distinct namespace
            // per account if the app ever holds more than one (account switching).
            secrets: KeychainSecretStore(namespace: "")
        )
        client = built
        return built
    }
}
